modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbMvPartitionStorage.java

```diff
@@ -0,0 +1,494 @@
+                    boolean found = true;
+
+                    // We should do it all the time. Here in particular it means one of two things:
+                    //  - partition is empty;
+                    //  - iterator exhausted all the data in partition.
+                    if (invalid(it)) {
+                        return false;
+                    }
+
+                    // At this point, seekKeyBuf should contain row id that's above the one we already scanned, but not greater than any
+                    // other row id in partition. When we start, row id is filled with zeroes. Value during the iteration is described later
+                    // in this code. Now let's descrie what we'll find, assuming that iterator found something:
+                    //  - if timestamp is null:
+                    //      - this seek will find the newest version of the next row in iterator. Exactly what we need.
+                    //  - if timestamp is not null:
+                    //      - suppose that seek key buffer has the following value: "| P0 | R0 | T0 |" (partition, row id, timestamp)
+                    //        and iterator finds something, let's denote it as "| P0 | R1 | T1 |" (partition must match). Again, there are
+                    //        few possibilities here:
+                    //          - R1 == R0, this means a match. By the rules of ordering we derive that T1 >= T0. Timestamps are stored in
+                    //            descending order, this means that we found exactly what's needed.
+                    //          - R1 > R0, this means that we found next row and T1 is either missing (pending row) or represents the latest
+                    //            version of the row. It doesn't matter in this case, because this row id will be reused to find its value
+                    //            at time T0. Additional "seek" will be required to do it.
+                    it.seek(seekKeyBuf.array());
+
+                    // Finish scan if nothing was found.
+                    if (invalid(it)) {
+                        return false;
+                    }
+
+                    // Read the actual key into a direct buffer.
+                    it.key(directBuffer);
+
+                    // To understand this condition please read that huge comment above.
+                    if (timestamp == null || !directBuffer.equals(seekKeyBufSliceWithoutTimestamp)) {
+                        // Copy actual row id into a "seekKeyBuf" buffer.
+                        GridUnsafe.copyMemory(
+                                null, GridUnsafe.bufferAddress(directBuffer) + ROW_ID_OFFSET,
+                                seekKeyBuf.array(), GridUnsafe.BYTE_ARR_OFF + ROW_ID_OFFSET,
+                                igniteRowIdSize
+                        );
+
+                        // Perform additional "seek" if timestamp is not null. Motivation for it is described in comments above.
```

Comment:
Maybe it would make sense to say that we are here because we ran out of the 'previous' row-related versions, so we must looks for a timestamp-matching version of the **next** row? I mean, when scanning, we need to find a version in each row. Reading the existing comments, I did not see this statement explicitly (or maybe I missed it).

Reply:
I don't know what to reply. Here we're looking for the matching version of "current" (known) row id, not a new (next) one. It's the last clause in that huge comment with many different options